forc-plugins/forc-client/src/util/aws.rs

```diff
@@ -0,0 +1,260 @@
+    let sig = sig.normalize_s().unwrap_or(sig);
+
+    // This is a hack to get the recovery id. The signature should be normalized
+    // before computing the recovery id, but aws kms doesn't support this, and
+    // instead always computes the recovery id from non-normalized signature.
+    // So instead the recovery id is determined by checking which variant matches
+    // the original public key.
+
+    let recid1 = RecoveryId::new(false, false);
+    let recid2 = RecoveryId::new(true, false);
+
+    let rec1 = VerifyingKey::recover_from_prehash(&*message, &sig, recid1);
+    let rec2 = VerifyingKey::recover_from_prehash(&*message, &sig, recid2);
+
+    let correct_public_key = k256::PublicKey::from_public_key_der(public_key_bytes)
+        .map_err(|_| anyhow::anyhow!("invalid DER public key from AWS KMS"))?
+        .into();
+
+    let recovery_id = if rec1.map(|r| r == correct_public_key).unwrap_or(false) {
+        recid1
+    } else if rec2.map(|r| r == correct_public_key).unwrap_or(false) {
+        recid2
+    } else {
+        anyhow::bail!("Invalid signature generated (reduced-x form coordinate)");
+    };
+
+    // Insert the recovery id into the signature
+    debug_assert!(
+        !recovery_id.is_x_reduced(),
+        "reduced-x form coordinates are caught by the if-else chain above"
+    );
+    let v = recovery_id.is_y_odd() as u8;
+    let mut signature = <[u8; 64]>::from(sig.to_bytes());
+    signature[32] = (v << 7) | (signature[32] & 0x7f);
+    Ok(fuel_crypto::Signature::from_bytes(signature))
+}
+
+impl AwsSigner {
+    pub async fn new(
+        kms: AwsClient,
+        key_id: String,
+        provider: Provider,
+    ) -> std::result::Result<Self, anyhow::Error> {
+        use k256::pkcs8::DecodePublicKey;
+
+        let resp = request_get_pubkey(kms.inner(), key_id.clone()).await?;
+        let public_key_bytes = decode_pubkey(&resp)?;
+        let k256_public_key = k256::PublicKey::from_public_key_der(&public_key_bytes)?;
+
+        let public_key = fuel_crypto::PublicKey::from(k256_public_key);
+        let hashed = public_key.hash();
+        let bech = Bech32Address::new(FUEL_BECH32_HRP, hashed);
+        Ok(Self {
+            kms,
+            key_id,
+            bech,
+            public_key_bytes,
+            provider,
+        })
+    }
+
+    /// Sign a digest with the key associated with a key ID.
+    pub async fn sign_message_with_key(
+        &self,
+        key_id: String,
+        message: Message,
+    ) -> std::result::Result<fuel_crypto::Signature, anyhow::Error> {
+        sign_with_kms(self.kms.inner(), &key_id, &self.public_key_bytes, message).await
+    }
+
+    /// Sign a digest with this signer's key
+    pub async fn sign_message(
+        &self,
+        message: Message,
+    ) -> std::result::Result<fuel_crypto::Signature, anyhow::Error> {
+        self.sign_message_with_key(self.key_id.clone(), message)
+            .await
+    }
```

Comment:
If we're going to call `sign_mesage()` many times, it might be worth just calling `sign_with_kms()` directly to avoid that `clone()`.

Reply:
Not for this change though, it's fine

Reply:
Yeah I am not really happy overall how this turned out tbh, once we are in clear, will come-back to this area. So I am adding this to never ending list of things to refactor :)